github/install.go

```diff
@@ -0,0 +1,232 @@
+		// - "-api.jar"
+		// - "-dev.jar"
+		// - "-dev-preshadow.jar"
+		// - "-sources.jar"
+		// In most cases, this will only match one asset.
+		// TODO: Hopefully.
+		regex := `^.+(?<!-api|-dev|-dev-preshadow|-sources)\.jar$`
+
+		// Check if the argument is a valid GitHub repository URL; if so, extract the slug from the URL.
+		// Otherwise, interpret the argument as a slug directly.
+		matches := GithubRegex.FindStringSubmatch(args[0])
+		if len(matches) == 2 {
+			slug = matches[1]
+		} else {
+			slug = args[0]
+		}
+
+		repo, err := fetchRepo(slug)
+
+		if err != nil {
+			fmt.Printf("Failed to add project: %s\n", err)
+			os.Exit(1)
+		}
+
+		if branchFlag != "" {
+			branch = branchFlag
+		}
+		if regexFlag != "" {
+			regex = regexFlag
+		}
+
+		err = installMod(repo, branch, regex, pack)
+		if err != nil {
+			fmt.Printf("Failed to add project: %s\n", err)
+			os.Exit(1)
+		}
+	},
+}
+
+func installMod(repo Repo, branch string, regex string, pack core.Pack) error {
+	latestRelease, err := getLatestRelease(repo.FullName, branch)
+	if err != nil {
+		return fmt.Errorf("failed to get latest release: %v", err)
+	}
+
+	return installRelease(repo, latestRelease, regex, pack)
+}
+
+func getLatestRelease(slug string, branch string) (Release, error) {
+	var releases []Release
+	var release Release
+
+	resp, err := ghDefaultClient.getReleases(slug)
+	if err != nil {
+		return release, err
+	}
+
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return release, err
+	}
+
+	err = json.Unmarshal(body, &releases)
+	if err != nil {
+		return release, err
+	}
+
+	if branch != "" {
+		for _, r := range releases {
+			if r.TargetCommitish == branch {
+				return r, nil
+			}
+		}
+		return release, fmt.Errorf("failed to find release for branch %v", branch)
+	}
+
+	return releases[0], nil
+}
+
+func installRelease(repo Repo, release Release, regex string, pack core.Pack) error {
+	expr := regexp2.MustCompile(regex, 0)
+
+	if len(release.Assets) == 0 {
+		return errors.New("release doesn't have any assets attached")
+	}
+
+	var files []Asset
+
+	for _, v := range release.Assets {
+		bl, _ := expr.MatchString(v.Name)
+		if bl {
+			files = append(files, v)
+		}
+	}
+
+	if len(files) == 0 {
+		return errors.New("release doesn't have any assets matching regex")
+	}
+
+	if len(files) > 1 {
+		// TODO: also print file names
+		return errors.New("release has more than one asset matching regex")
+	}
+
+	file := files[0]
+
+	// Install the file
+	fmt.Printf("Installing %s from release %s\n", file.Name, release.TagName)
+	index, err := pack.LoadIndex()
+	if err != nil {
+		return err
+	}
+
+	updateMap := make(map[string]map[string]interface{})
+
+	updateMap["github"], err = ghUpdateData{
+		Slug:   repo.FullName,
+		Tag:    release.TagName,
+		Branch: release.TargetCommitish, // TODO: if no branch is specified by the user, we shouldn't record it - in order to remain branch-agnostic in getLatestRelease()
```

Comment:
I do think some of the behaviour of using a branch by default is useful! On Indium for example, it'd use 1.21.x/stable and only update to 1.21 versions, instead of updating to a 1.22 release. I don't know how widely this applies though, and opt-in instead of opt-out to branch filtering probably makes more sense.

Reply:
Hmmm fair! I do worry that regex will be a bit complicated for most users... tho that could be solved with good documentation?